iroh-gateway/src/handler.rs

```diff
@@ -0,0 +1,308 @@
+use axum::{
+    body::{self, Body, BoxBody},
+    error_handling::HandleErrorLayer,
+    extract::{Extension, Path, Query},
+    http::StatusCode,
+    response::IntoResponse,
+    routing::get,
+    BoxError, Router,
+};
+use serde::Deserialize;
+use std::{borrow::Cow, collections::HashMap, time::Duration};
+use tower::ServiceBuilder;
+
+use crate::client::Client;
+use crate::config::Config;
+use crate::error::GatewayError;
+use crate::response::{GatewayResponse, ResponseFormat};
+
+pub struct Handler {
+    pub config: Config,
+    client: Client,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct GetParams {
+    format: Option<String>,
+    filename: Option<String>,
+    download: Option<String>,
+}
+
+impl Handler {
+    pub fn new(config: Config) -> Self {
+        Self {
+            config,
+            client: Client::new(),
+        }
+    }
+
+    async fn get_ipfs(
+        Extension(config): Extension<Config>,
+        Extension(client): Extension<Client>,
+        Path(params): Path<HashMap<String, String>>,
+        Query(query_params): Query<GetParams>,
+    ) -> Result<GatewayResponse, GatewayError> {
+        let cid = params.get("cid").unwrap();
+        let cpath = "".to_string();
+        let cpath = params.get("cpath").unwrap_or(&cpath);
+        let full_content_path = format!("/ipfs/{}{}", cid, cpath);
+        println!("cid: {}", cid);
+        println!("cpath: {}", cpath);
+        println!("fullpath: {}", full_content_path);
+
+        let format = query_params.format.unwrap_or("".to_string());
+        let format = Handler::response_format(&format);
+        let format = match format {
+            Ok(f) => f,
+            Err(error) => {
+                let msg = format!("{}", error);
+                return Handler::error(StatusCode::BAD_REQUEST, &msg);
+            }
+        };
+        let query_file_name = query_params.filename.unwrap_or("".to_string());
+        let download = query_params.download.unwrap_or("".to_string()) == "true";
+
+        let mut headers = Handler::format_headers(&format);
+        headers.insert("X-Ipfs-Path".to_string(), full_content_path.clone());
```

Comment:
I would try and make some constants for all the different headers we use, maybe in its own file